### PR TITLE
Port to GUI UI library

### PR DESCRIPTION
**Pull request by scartill**
- New library for UI
- Rearranged the GUI
- Bonus: natively works under Windows (tested with Windows 10)
- No logic changed (in theory)

### CODE REVIEW

**Comment by iegomez** [comment]
Holy PRs, @scartill, that's a big one! 👏 

Let me check it out locally to confirm it build and runs nicely (I take it you're working on Windows, I do on Linux but also have a Mac machine for work), and go through the code, but kudos!

**Review by iegomez — CHANGES_REQUESTED** [review]
Just a quick feedback before I dive into it:

- In Mac it builds without issues but in Linux (Mint 19.3) I was missing the `xkbcommon-x11` package, so I had tun run this:
```
sudo apt-get install libxkbcommon-x11-0 libxkbcommon-x11-dev
```
- In both cases `tab` doesn't switch between inputs (maybe a `gioui` limitation?), and I'm also not able to select MQTT username nor password inputs manually.
- The layout looks pretty messy and condensed in both Mac and Linux. It may have to do with both machines having HiDPI displays, not sure at this moment. I'm attaching an image of running it on Mint, but it looked the same in Mac:

Screenshot from 2020-07-02 20-29-26

**Comment by scartill** [comment]
@iegomez, many thanks for taking a look!

Sorry for a huge PR, but that was the only consistent way 

* Yeah, that looks like a mess. That have to do with screen resolution, I believe. I've tested user the same Linux Mint with no trouble. Will try to play with resolution and see what happens.
* As for dependency, that is indeed derived from GIO UI, will add it to documentation.
* Tabs, I'm afraid, are GIO UI limitations, but that library is under heavy development - many things are upcoming.

For the reference, attaching the screenshot for Mint, pretty much the same is under Windows (no Mac to test on, sorry :)

image

**Comment by iegomez** [comment]
Ah, indeed that looks pretty nice, so probably it's the HiDPI scaling that's messing it.

One more thing: when we discussed it in the issue, I thought you were gonna add a separate GUI using `gio ui` instead of replacing the `dear imgui` one, but maybe I misunderstood. I'm open for discussion though: as I've mentioned on other occasions, I'm pretty tight with work and trying to get `mosquitto-go-auth` to 1.0, so it makes much more sense for you to be making the big decisions and me backing off a bit and just reviewing stuff.

**Comment by scartill** [comment]
I've changed layout dimensions to device-independent pixel, which may fixed HiDP issue (in theory, cannot test, pardon), and made some minor layout changes.

I'm sorry if I've caused any misunderstanding with `dear imgui`, but as to keep it - that may be doable, but requires even more major overhaul. I'm not sure there is a need for dual interface.

**Comment by iegomez** [comment]
It looks a bit better, input sizes and the console definitely saw a great improvement, so kudos!

Sadly, side buttons are still messed:
Screenshot from 2020-07-11 19-46-59

As for keeping `dear imgui` interface, don't worry: if I find the time I'll open a different branch with it, maybe see if the original issue is fixed or I can do anything about it, and then we can merge it to have both. We'll see. :wink:

**Comment by scartill** [comment]
Made a change that would (probably :) fix the button panel.

**Comment by iegomez** [comment]
Yesterday I messed my local machine with a bad Linux Mint upgrade, so today I reconfigured but I'm running into some issue trying to build:
```
Package wayland-egl was not found in the pkg-config search path.
Perhaps you should add the directory containing `wayland-egl.pc'
to the PKG_CONFIG_PATH environment variable
No package 'wayland-egl' found
Package wayland-client was not found in the pkg-config search path.
Perhaps you should add the directory containing `wayland-client.pc'
to the PKG_CONFIG_PATH environment variable
No package 'wayland-client' found
Package wayland-cursor was not found in the pkg-config search path.
Perhaps you should add the directory containing `wayland-cursor.pc'
to the PKG_CONFIG_PATH environment variable
No package 'wayland-cursor' found
Package x11-xcb was not found in the pkg-config search path.
Perhaps you should add the directory containing `x11-xcb.pc'
to the PKG_CONFIG_PATH environment variable
No package 'x11-xcb' found
```
So let me fix that and then I'll check it.

**Comment by iegomez** [comment]
@scartill Just built it in the Mac and side buttons are looking good!

**Comment by scartill** [comment]
That's great, thank you!

**Comment by scartill** [comment]
Hi @iegomez,

Should I amend these changes further? What are your thoughts?

**Comment by iegomez** [comment]
Hey @scartill!

Let me take a final look today after work and then you should be all set.

File: config.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+	"github.com/brocaar/lorawan"
+	log "github.com/sirupsen/logrus"
+
```
Comment (iegomez) [inline config.go]: Could you remove the empty line? It should also sort external imports.

File: config.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	config   *tomlConfig
+)
+
+// Configuration files loading and saving.
+var (
+	openFile     bool
+	files        []os.FileInfo
+	saveFile     bool
+	saveFilename string
+	mwOpen       = true
```
Comment (iegomez) [inline config.go]: What's the use for this? I see no other mentions in the project.

File: config.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		}
+
+		cBand := band{}
+
+		cDr := dataRate{}
+
+		cRx := rxInfo{}
+
+		cPl := rawPayload{
+			MaxExecTime: 100,
```
Comment (iegomez) [inline config.go]: This could be a a `defaultMaxExecTime` package constant.

File: config.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	saveFilename string
+	mwOpen       = true
+)
+
+func importConf() {
+
+	//When config hasn't been initialized we need to provide fresh zero instances with some defaults.
+	//Decoding the conf file will override any present option.
+	if config == nil {
+		cMqtt := mqtt{}
```
Comment (iegomez) [inline config.go]: I know this is my fault, but it'd be nice if we just throw all these declarations as literals in `config` assignment.

File: config.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+		log.Println(err)
+		return
+	}
+
+	l, err := log.ParseLevel(config.LogLevel)
+	if err != nil {
+		log.SetLevel(log.InfoLevel)
+	} else {
+		log.SetLevel(l)
+	}
```
Comment (iegomez) [inline config.go]: Again, this is on me, but we could change it to:
```
log.SetLevel(log.InfoLevel)
if l, err := log.ParseLevel(config.LogLevel); err != nil {
    log.SetLevel(l)
}
```
Even better, we could have a `map[string]struct{}` with allowed levels, but that could be done later.

File: data.go [hunk 6]
```diff
@@ -4,12 +4,19 @@ import (
 	"encoding/json"
 	"fmt"
 	"reflect"
+	c "strconv"
```
Comment (iegomez) [inline data.go]: I'm not a fan of renaming packages unless there's name collisions, it makes it less explicit.

File: data.go [hunk 7]
```diff
@@ -4,12 +4,19 @@ import (
...
 	"time"
 
-	"github.com/inkyblackness/imgui-go"
 	"github.com/pkg/errors"
 	"github.com/robertkrimen/otto"
 	log "github.com/sirupsen/logrus"
+
+	l "gioui.org/layout"
```
Comment (iegomez) [inline data.go:10]: Same as above, and please remove the trailing empty line to get imports sorted.

File: forwarder.go [hunk 8]
```diff
@@ -3,10 +3,15 @@ package main
 import (
 	"strconv"
```
Comment (iegomez) [inline forwarder.go:5]: Empty line here and other imports.

File: forwarder.go [hunk 9]
```diff
@@ -3,10 +3,15 @@ package main
...
 
-	"github.com/inkyblackness/imgui-go"
 	log "github.com/sirupsen/logrus"
 
 	"github.com/iegomez/lds/lds"
+
+	l "gioui.org/layout"
```
Comment (iegomez) [inline forwarder.go:6]: See comments above.

File: giox_helpers.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"strconv"
+
+	l "gioui.org/layout"
```
Comment (iegomez) [inline giox_helpers.go:6]: Once again.

File: lora.go [hunk 11]
```diff
@@ -4,7 +4,13 @@ import (
 	"strconv"
 
 	lwband "github.com/brocaar/lorawan/band"
-	"github.com/inkyblackness/imgui-go"
+
+	l "gioui.org/layout"
```
Comment (iegomez) [inline lora.go:6]: Same old and empty lines.

File: mqtt.go [hunk 12]
```diff
@@ -5,8 +5,13 @@ import (
 	"time"
 
 	paho "github.com/eclipse/paho.mqtt.golang"
-	"github.com/inkyblackness/imgui-go"
 	log "github.com/sirupsen/logrus"
+
+	l "gioui.org/layout"
```
Comment (iegomez) [inline mqtt.go:7]: :roll_eyes:

File: output.go [hunk 13]
```diff
@@ -5,8 +5,12 @@ import (
 	"os"
 	"time"
 
-	"github.com/inkyblackness/imgui-go"
 	log "github.com/sirupsen/logrus"
+
+	l "gioui.org/layout"
```
Comment (iegomez) [inline output.go:8]: :eyes:

**Review by iegomez — CHANGES_REQUESTED** [review]
Mostly nitpicks about style, I don't know `gioui` to do a helpful review in that regard.

Also, could you add changes to the docs to build the app?

**Comment by scartill** [comment]
Hi @iegomez ,

I've tried to address all the issues, and updated the doc.

Regards,
Boris

**Review by iegomez — APPROVED** [review]

**Comment by iegomez** [comment]
Looks good! :rocket:

**Comment by scartill** [comment]
Merged, thank you!

**Pull request merged**